x/qgb/README.md

```diff
@@ -1,50 +1,320 @@
+
+After a range update, the [`handler`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L37-L82) tries to [generate](https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/abci.go#L51-L81) any needed data commitment to catchup to the current height. An example of how these ranges are created can be found in [tests](https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/abci_test.go#L412-L520). Thus, in the same height, multiple data commitments can be generated and added to the store.
+
+### Pruning
+
+The third action done during the QGB [`EndBlock`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L28-L35) step is pruning.
+
+The QGB state machine prunes old attestations up to the specified [`AttestationExpiryTime`](https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/abci.go#L22-L25), which is currently set to 3 weeks, matching the consensus unbonding time.
+
+So, on every block height, the state machine [checks](https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/abci.go#L140-L157) whether there are any [`expired`]((https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/abci.go#L22-L25)) attestations. Then, it starts [pruning](https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/abci.go#L161-L182) via calling the [`DeleteAttestation(...)`](https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/keeper/keeper_attestation.go#L128-L139) method. Then, it [`prints`](https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/abci.go#L186-L194) a log message specifying the number of pruned attestations.
+
+If the all the attestations in store are expired, which is an edge case that should never occur, the QGB state machine [doesn't prune](https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/abci.go#L161) the latest attestation.
+
+### Hooks
+
+#### Validator unbonding hook
+
+When a validator starts unbonding, a [hook](https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/keeper/hooks.go#L23-L34) is executed that [sets](https://github.com/celestiaorg/celestia-app/blob/0629c757ef35a24187a8d7a4c706c7cdc894c8b6/x/qgb/keeper/hooks.go#LL33C2-L33C2) the `LatestUnBondingBlockHeight` to the current block height. This allows creating a new valset that removes that validator from the valset members so that he doesn't need to sign attestations afterwards.
```

Comment:
Is there a threshold for how much they should unbond before this gets triggered?

Reply:
There is no threshold, the time a validator starts unbonding, this gets triggered

Reply:
So if a validator unbonds a single TIA token this will trigger? (Even though there would be no noticeable change to the validator set)

Reply:
In this case yes, because now he has lower power in the network, and it might make a difference when calculating the 2/3 relaying threshold

Reply:
But why is a validator unbonding a single TIA token different to a delegator unbonding a single TIA token from the same validator. They both have the same voting power difference. I don't see we can't just use the significant voting power threshold as the sole trigger for a valset attestation.

Couldn't this also be some form of spam. i.e. every height I unbond 1 TIA, meaning that an attestation is made every height.

Reply:
> I don't see we can't just use the significant voting power threshold as the sole trigger for a valset attestation.

We could but we won't be catching the case where a validator unbonds all his tokens, which are `4.9999999` of the power, and he would be able to double sign afterward without us slashing him.

> Couldn't this also be some form of spam. i.e. every height I unbond 1 TIA, meaning that an attestation is made every height.

They would also be paying the transaction fees. And we have pruning, so it would be around 500mb if they create a valset every block for 3 weeks for 15sec block time. 500mb of state that is not really manipulated inside the state machine that much is okey, I guess

Reply:
> We could but we won't be catching the case where a validator unbonds all his tokens, which are 4.9999999 of the power, and he would be able to double sign afterward without us slashing him.

Don't you require 1/3 to double sign in order to genuinely equivocate (I'm just assuming light client verification security. I actually don't know the specifics)

Also it will take 21 days for the validator to truly unbond (or is the unbond hook for when their tokens are unlocked).

Reply:
> Don't you require 1/3 to double sign in order to genuinely equivocate

Yes, but even if they will be unable to halt the QGB, we still would want to slash any bad behavior

> Also it will take 21 days for the validator to truly unbond (or is the unbond hook for when their tokens are unlocked).

But, if after they start unbonding their tokens, we don't create a valset. Let's imagine a scenario where during that 21 days, no change happens to the valset. Then, day 22 we will still have the relayer getting the signature of that 
 malicious orchestrator, his tokens will be unlocked, and he will be able to submit any signature and will get relayed to the QGB. Of course, we're implementing the relayer to be smarter and not relay it, but also we wouldn't want this behaviour to go unpunished

Reply:
> Yes, but even if they will be unable to halt the QGB, we still would want to slash any bad behavior

I'm usually a punishment minimalist. So long as there's not a 1/3 validator set change, there's nothing a malicious validator can do. I think 5% is a good threshold in this case.